File: neet_code_150_cpp/backtracking/combination_sum.cpp
#include <iostream>
#include <vector>
#include <numeric>

// 模範解答
class SolutionAns {
public:
    std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
        std::vector<int> curr;
        std::vector<std::vector<int>> res;
        helper(candidates, target, 0, curr, res);
        return res;
    }

private:
    void helper(
            std::vector<int>& cands, 
            int target, 
            int i, 
            std::vector<int>& curr,  
            std::vector<std::vector<int>>& res)
    {
        // iがcandsのサイズを超える、もしくはtarget < 0なら終了
        if (i >= cands.size() || target < 0) {
            return;
        }
        // target が0になったら終了
        if (target == 0) {
            res.push_back(curr);
            return;
        }

        // cand[i]を選ぶ場合
        // i番目をcurrにプッシュ
        curr.push_back(cands[i]);
        helper(cands, target - cands[i], i, curr, res);

        // cand[i]を選ばない場合
        // 末尾の要素を削除
        curr.pop_back();
        helper(cands, target, i + 1, curr, res);
    }
};

// Rustの模範解答より
class SolutionRustAns {
    public:
        std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
            std::vector<int> current;
            std::vector<std::vector<int>> result;
            dfs(candidates, target, 0, result, current);

            return result;
        }

    private:
        void dfs(
                std::vector<int>& candidates,
                int target,
                int idx,
                std::vector<std::vector<int>>& result,
                std::vector<int>& current
                ) {
            // currentの和がtargetに等しければ終了
            auto sum = std::accumulate(current.begin(), current.end(), 0); 
            if(sum == target) {
                result.push_back(current);
                return;
                // sum がtargetの値を超えたら終了
            } else if(sum > target) {
                return;
            }

            // currentの更新
            for(int i = idx; i < candidates.size(); ++i) {
                current.push_back(candidates[i]);
                // i番目以上の要素を選ぶ
                dfs(candidates, target, i, result, current);
                current.pop_back();
            }
        }
};

int main(void) {
    auto case_1 = std::pair(std::vector{2, 3, 6, 7}, 7);
    // => {{2,2,3},{7}}
    auto case_2 = std::pair(std::vector{2, 3, 5}, 8);
    // => {{2,2,2,2}, {2,3,3}, {3,5}}
    auto case_3 = std::pair(std::vector{2}, 1);
    // => {}
    auto case_4 = std::pair(std::vector{2, 3, 6, 7}, 7);
    // => {{2, 3, 6, 7}, {7}}

    SolutionAns s_ans;
    auto res_1 = s_ans.combinationSum(case_1.first, case_1.second);
    auto res_2 = s_ans.combinationSum(case_2.first, case_2.second);
    auto res_3 = s_ans.combinationSum(case_3.first, case_3.second);
    auto res_4 = s_ans.combinationSum(case_4.first, case_4.second);

    std::cout << "case_1" << std::endl;
    for(const auto &a: res_1) {
        for(const auto &b: a){
            std::cout << b << " ";
        }
        std::cout << std::endl;
    }

    std::cout << "case_2" << std::endl;
    for(const auto &a: res_2) {
        for(const auto &b: a){
            std::cout << b << " ";
        }
        std::cout << std::endl;
    }

    std::cout << "case_3" << std::endl;
    for(const auto &a: res_3) {
        for(const auto &b: a){
            std::cout << b << " ";
        }
        std::cout << std::endl;
    }

    std::cout << "case_4" << std::endl;
    for(const auto &a: res_4) {
        for(const auto &b: a){
            std::cout << b << " ";
        }
        std::cout << std::endl;
    }

    SolutionRustAns s_rust_ans;
    auto res_rust_1 = s_rust_ans.combinationSum(case_1.first, case_1.second);
    auto res_rust_2 = s_rust_ans.combinationSum(case_2.first, case_2.second);
    auto res_rust_3 = s_rust_ans.combinationSum(case_3.first, case_3.second);
    auto res_rust_4 = s_rust_ans.combinationSum(case_4.first, case_4.second);

    std::cout << "case_1" << std::endl;
    for(const auto &a: res_1) {
        for(const auto &b: a){
            std::cout << b << " ";
        }
        std::cout << std::endl;
    }

    std::cout << "case_2" << std::endl;
    for(const auto &a: res_2) {
        for(const auto &b: a){
            std::cout << b << " ";
        }
        std::cout << std::endl;
    }

    std::cout << "case_3" << std::endl;
    for(const auto &a: res_3) {
        for(const auto &b: a){
            std::cout << b << " ";
        }
        std::cout << std::endl;
    }

    std::cout << "case_4" << std::endl;
    for(const auto &a: res_4) {
        for(const auto &b: a){
            std::cout << b << " ";
        }
        std::cout << std::endl;
    }
}

File: neet_code_150/src/backtracking/combination_sum.rs
// 解けなかった
struct Solution {}
impl Solution {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut tmp = 0;
        let mut idx = 0;
        while tmp < target {}

        vec![vec![1]]
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let (mut result, mut curr) = (vec![], vec![]);
        Self::dfs(&candidates, target, &mut result, &mut curr);

        result
    }

    fn dfs(candidates: &[i32], target: i32, result: &mut Vec<Vec<i32>>, curr: &mut Vec<i32>) {
        // 現在のcurrの和がtargetになるなら終了
        let sum: i32 = curr.iter().sum();
        if sum == target {
            result.push(curr.to_owned());
            return;
        // sumがtargetを超えたら終了
        } else if sum > target {
            return;
        }

        // currの更新
        for (i, &c) in candidates.iter().enumerate() {
            curr.push(c);
            Self::dfs(&candidates[i..], target, result, curr);
            curr.pop();
        }
    }
}

// C++の模範解答より
struct SolutionCppAns {}
impl SolutionCppAns {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut current = vec![];
        let mut result = vec![];
        Self::helper(&candidates, target, 0, &mut current, &mut result);

        result
    }

    fn helper(
        candidates: &Vec<i32>,
        target: i32,
        idx: usize,
        current: &mut Vec<i32>,
        result: &mut Vec<Vec<i32>>,
    ) {
        if idx >= candidates.len() || target < 0 {
            return;
        }

        if target == 0 {
            result.push(current.clone());
            return;
        }

        //idx番目を選ぶ場合
        //idx番目をcurrentにpush
        current.push(candidates[idx]);
        Self::helper(candidates, target - candidates[idx], idx, current, result);

        // idx番目を選ばない場合
        // 末尾の要素を削除
        current.pop();
        Self::helper(candidates, target, idx + 1, current, result);
    }
}

fn main() {
    let case_1 = (vec![2, 3, 6, 7], 7);
    // => [[2,2,3],[7]]
    let case_2 = (vec![2, 3, 5], 8);
    // => [[2,2,2,2], [2,3,3], [3,5]]
    let case_3 = (vec![2], 1);
    // => [[]]
    let case_4 = (vec![2, 3, 6, 7], 7);
    // => [[2,2,3],[7]]

    println!(
        "case_1: {:?}",
        SolutionAns::combination_sum(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {:?}",
        SolutionAns::combination_sum(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {:?}",
        SolutionAns::combination_sum(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {:?}",
        SolutionAns::combination_sum(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_1: {:?}",
        SolutionCppAns::combination_sum(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {:?}",
        SolutionCppAns::combination_sum(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {:?}",
        SolutionCppAns::combination_sum(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {:?}",
        SolutionCppAns::combination_sum(case_4.0.clone(), case_4.1)
    );
}
